src/rgw/rgw_cr_rest.cc

```diff
@@ -0,0 +1,351 @@
+        }
+        bl.splice(0, max, &extra_data);
+        bl_len -= max;
+        got_all_extra_data = extra_data.length() == extra_data_len;
+      }
+
+      data.append(bl);
+    }
+
+#define GET_DATA_WINDOW_SIZE 2 * 1024 * 1024
+    uint64_t data_len = data.length();
+    if (data_len >= GET_DATA_WINDOW_SIZE) {
+      env->manager->io_complete(cr, io_id);
+    }
+    if (data_len >= 2 * GET_DATA_WINDOW_SIZE) {
+      *pause = true;
+      paused = true;
+    }
+    return 0;
+  }
+
+  void claim_data(bufferlist *dest, uint64_t max) {
+    bool need_to_unpause = false;
+
+    {
+      Mutex::Locker l(lock);
+
+      if (data.length() == 0) {
+        return;
+      }
+
+      if (data.length() < max) {
+        max = data.length();
+      }
+
+      data.splice(0, max, dest);
+      need_to_unpause = (paused && data.length() <= GET_DATA_WINDOW_SIZE);
+    }
+
+    if (need_to_unpause) {
+      req->unpause_receive();
+    }
+  }
+
+  bufferlist& get_extra_data() {
+    return extra_data;
+  }
+
+  bool has_data() {
+    return (data.length() > 0);
+  }
+
+  bool has_all_extra_data() {
+    return got_all_extra_data;
+  }
+};
+
+
+RGWStreamReadHTTPResourceCRF::~RGWStreamReadHTTPResourceCRF()
+{
+  delete in_cb;
+}
+
+int RGWStreamReadHTTPResourceCRF::init()
+{
+  env->stack->init_new_io(req);
+
+  in_cb = new RGWCRHTTPGetDataCB(env, caller, req);
+
+  int r = http_manager->add_request(req);
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+int RGWStreamWriteHTTPResourceCRF::send()
+{
+  env->stack->init_new_io(req);
+
+  req->set_write_drain_cb(&write_drain_notify_cb);
+
+  int r = http_manager->add_request(req);
+  if (r < 0) {
+    return r;
+  }
+
+  return 0;
+}
+
+bool RGWStreamReadHTTPResourceCRF::has_attrs()
+{
+  return got_attrs;
+}
+
+void RGWStreamReadHTTPResourceCRF::get_attrs(std::map<string, string> *attrs)
+{
+  req->get_out_headers(attrs);
+}
+
+int RGWStreamReadHTTPResourceCRF::decode_rest_obj(map<string, string>& headers, bufferlist& extra_data, rgw_rest_obj *info) {
+  /* basic generic implementation */
+  for (auto header : headers) {
+    const string& val = header.second;
+
+    rest_obj.attrs[header.first] = val;
+  }
+
+  return 0;
+}
+
+int RGWStreamReadHTTPResourceCRF::read(bufferlist *out, uint64_t max_size, bool *io_pending)
+{
+    reenter(&read_state) {
+    io_read_mask = req->get_io_id(RGWHTTPClient::HTTPCLIENT_IO_READ | RGWHTTPClient::HTTPCLIENT_IO_CONTROL);
+    while (!req->is_done() ||
+           in_cb->has_data()) {
+      *io_pending = true;
+      if (!in_cb->has_data()) {
+        yield caller->io_block(0, io_read_mask);
+      }
+      got_attrs = true;
+      if (need_extra_data() && !got_extra_data) {
+        if (!in_cb->has_all_extra_data()) {
+          continue;
+        }
+        extra_data.claim_append(in_cb->get_extra_data());
+        map<string, string> attrs;
+        req->get_out_headers(&attrs);
+        int ret = decode_rest_obj(attrs, extra_data, &rest_obj);
```

Comment:
since `attrs` isn't used after this call, consider passing it as `std::move(attrs)` so that to `decode_rest_obj()` doesn't have to copy the `map<string,string>`